### DISCUSSION TITLE
HELP: supabase OAuth seems confusing to me

### DISCUSSION THREAD

**Discussion by kaf-lamed-beyt** [supabase/supabase#21203] [Questions] [answered]
So, I've been trying to setup OAuth for a while now, to no avail. I've tried GitHub and Twitter OAuth, and the process remains the same. The docs themselves do not do complete justice to the explanation. It just includes the following snippet: 

```js
const { data, error } = await supabase.auth.signInWithOAuth({
  provider: "github",
});
```

The information on redirectUrls isn't detailed. Say, for example, I try using the snippet above with the twitter provider enabled and all my settings are  accounted for. The auth flow after authorizing the OAuth app in the github dialog takes me back to my home page, and nothing happens. I have a `<AuthProvider />` component that listens for `supabase.auth.onAuthStateChange()` and redirects to the dashboard if the user is authenticated or if a session is available. 

But no session is available, with this OAuth flow. I don't know what I'm doing wrong. 

I even went further to include a `redirectTo` property in the OAuth sign-in method like so: 

```js
const { data, error } = await supabase.auth.signInWithOAuth({
   provider: "github",
   options: {
      redirectTo: "http://localhost:5173/oauth",
   },
});
```

So I can read the `access_token` from the URL segment on my `/oauth` route and probably exchange it for a session. But, I am still lost. Please any help would be much appreciated. I'm quite new to Supabase. 

For context, this is what my `/oauth` component looks like:

```js
import React from "react";
import { setCookie } from "cookies-next";
import { Center, Text } from "@chakra-ui/react";
import { useLocation, useNavigate } from "react-router-dom";
import { useToastContext } from "@hooks/toast";
import { authCookieOptions } from "@utils/misc";
import { useAuthContext } from "@hooks/auth";

export const Oauth = () => {
  const location = useLocation();
  const navigate = useNavigate();
  const { openToast } = useToastContext();
  const { authenticator } = useAuthContext();

  React.useEffect(() => {
    const params = new URLSearchParams(location.hash.slice(1));
    const accessToken = params.get("access_token");

    try {
      if (accessToken) {
        setCookie("_gat", accessToken, {
          ...authCookieOptions,
        });

        openToast("Logged in successfully!", "success");
        navigate("/dashboard");
      } else {
        navigate("/signin");
        openToast("something went wrong! Try again.", "error");
      }
    } catch (error) {
      openToast(`${error}`, "error");
    }
  }, [authenticator, location, navigate, openToast]);

  return (
    <Center height="100vh">
      <Text>Please wait, while we take you to your dashboard...</Text>
    </Center>
  );
};
```

**Comment by charislam** [comment]
If you have an access token in the URL, that's already your session, and there's no further need to exchange it. What _should_ be happening is that Supabase will automatically parse the access token from the URL and save the session to local storage. It appears that isn't occurring, so first question to help troubleshoot:

It looks like you're using Next.js. Are you using the App Router or Pages Router?

**Comment by kaf-lamed-beyt** [comment, marked as answer]
No, I'm not using Next.js.  

And I found a workaround. I don't know if it is the right way though. But, it helped me get what I wanted. 

On the /oauth route, where I'm redirected to. I had to exchange the access token for a user and redirect to my dashboard if the process is successful. 

```ts
const { data, error } = await supabase.auth.getUser(accessToken) 

If (!error) {
  updateUserState(data?.user)
  navigate("to/whereever")
}
```

**Reply by kaf-lamed-beyt** [reply]
I wrote about how I addressed it in detail, [here](https://www.meje.dev/blog/supabase-oauth-in-react-apps) in case anyone stumbles upon this.

**Reply by Chippd** [reply]
Stumbled across this as I hit an issue recently with sessions not being established properly with either the OAuth flow or email magic link flows. 
For me, the issue was specific to certain pages where I would initiate the auth flow (my app has an auth modal that I pop up on whatever page the user wants/needs to log in on). 
On some of these pages, I had query parameters in the URL. when setting the `redirectTo` value in the `options` of the signInWithOAuth call, I was just pulling in the current URL via `window.location.href` so that my users would be brought back to the same page they were on. When that page contained query parameters, my users would get dumped back on the homepage of the app and no session would be initiated. 
When I strip the query params from the redirectTo value, all works properly. 

Looking at your blog post, I believe all that extra logic you wrote shouldn't be necessary, I simply call getSession() and I have the user's session when they get redirected to my app. I've also added a `onAuthStateChange()` listener in case that getSession call doesn't work for whatever reason. 

Hope this helps you, someone else, or me in the future.

**Reply by kaf-lamed-beyt** [reply]
Aha! True. I recently also learned about `getSession` and I've been using for a while now. 

I'll try modifying the article when I'm less occupied. 

Thank you for this feedback!
